### PR TITLE
WIP: Event processing should handle reorgs / undo

### PR DESCRIPTION
**Pull request by jflatow**

### CODE REVIEW

File: ethereum-client/src/events.rs [hunk 1]
```diff
@@ -312,54 +312,58 @@ pub fn decode_event(topics: Vec<String>, data: String) -> Result<EthereumEvent,
...
         .collect::<Result<Vec<ethabi::Hash>, _>>()?;
-    let topic_hash = topic_hashes.first().ok_or(EventError::InvalidTopic)?;
-    if *topic_hash == *LOCK_EVENT_TOPIC {
-        let log: ethabi::Log = LOCK_EVENT
-            .parse_log(ethabi::RawLog {
-                topics: topic_hashes,
-                data: decode_hex(&data).ok_or(EventError::InvalidHex)?,
-            })
-            .map_err(|_| EventError::ErrorParsingLog)?;
+    match topic_hashes.first().ok_or(EventError::InvalidTopic)? {
```
Comment (hayesgm) [inline ethereum-client/src/events.rs:315]: Cool "match -> cond" hack

File: pallets/cash/src/internal/events.rs [hunk 2]
```diff
@@ -1,182 +1,416 @@
...
+    last_block: &ChainBlock,
+    event_queue: &mut ChainBlockEvents,
+) -> Result<(), Reason> {
+    use crate::types::{Quantity, USDQuantity, USD};
+    const QUOTA: USDQuantity = Quantity::from_nominal("10000", USD); // XXX value? per chain?
+    let mut available = QUOTA;
+    event_queue.retain(|event| {
+        // note that we panic if we cannot compute the risk adjusted value,
+        //  there's not an obviously more reasonable thing to do right now
+        let value = risk_adjusted_value::<T>(event, last_block.number()).unwrap();
```
Comment (hayesgm) [inline pallets/cash/src/internal/events.rs:135]: I might consider returning with no action or ignoring the event versus panicing. Panics basically prevent any progress on the chain until every node hard-forks, right?

File: pallets/cash/src/internal/events.rs [hunk 3]
```diff
@@ -1,182 +1,416 @@
...
+pub fn formulate_reorg<T: Config>(
+    chain_id: ChainId,
+    last_block: &ChainBlock,
+    true_block: &ChainBlock,
+) -> Result<ChainReorg, Reason> {
+    // XXX TODO whatever we submit we must store first so we can formulate reorgs
+    //  and worry about pruning history independently / later
+    const CHUNK: u32 = 10;
+    let mut reverse_blocks: Vec<ChainBlock> = vec![];
+    let mut drawrof_blocks: Vec<ChainBlock> = vec![];
```
Comment (hayesgm) [inline pallets/cash/src/internal/events.rs:212]: What does `drawrof` mean?

File: pallets/cash/src/internal/events.rs [hunk 4]
```diff
@@ -1,182 +1,416 @@
...
-            signer,
-            validators
-        );
-        return Err(Reason::UnknownValidator)?;
+    let validator_set = get_validator_set::<T>()?;
+    let validator = recover_validator::<T>(&blocks.encode(), signature)?;
+    let chain_id = blocks.chain_id();
+    let mut last_block = LastProcessedBlock::get(chain_id).ok_or(Reason::NoSuchBlock)?;
+    let mut pending_blocks = PendingChainBlocks::get(chain_id);
+    let mut event_queue = IngressionQueue::get(chain_id).ok_or(Reason::NoSuchQueue)?;
```
Comment (hayesgm) [inline pallets/cash/src/internal/events.rs]: Where is the queue initialized?

File: pallets/cash/src/internal/events.rs [hunk 5]
```diff
@@ -1,182 +1,416 @@
...
-    log!("Fetching events starting from block {:?}", from_block);
+            EthereumEvent::LockCash { principal, .. } => {
+                let quantity = get_cash_quantity::<T>(CashPrincipalAmount(*principal))?;
+                let usd_quantity = get_value::<T>(quantity)?;
+                Ok(usd_quantity.decay(elapsed_blocks)?)
+            }
+
+            EthereumEvent::ExecuteProposal { .. } => {
+                // XXX TODO choose some large value?
+                const LARGE_QTY: USDQuantity = Quantity::from_nominal("1000000000000", USD);
```
Comment (jflatow) [inline pallets/cash/src/internal/events.rs]: Shall we introduce a param for this?

**Review by hayesgm — COMMENTED** [review]
looking super 🔥🔥🔥

File: pallets/cash/src/internal/events.rs (hunk 4 above)
Comment (jflatow) [reply pallets/cash/src/internal/events.rs:275]: From `Default` (empty)

Comment (jflatow) [reply pallets/cash/src/internal/events.rs]: Good ❓ Should we default to empty?

File: pallets/cash/src/internal/events.rs (hunk 2 above)
Comment (jflatow) [reply pallets/cash/src/internal/events.rs:135]: Doing that inside the retain creates a lot of extra complexity, and there's no real reason we can't compute this though

File: pallets/cash/src/internal/events.rs (hunk 3 above)
Comment (jflatow) [reply pallets/cash/src/internal/events.rs:212]: Its the `forward_blocks` in reverse order (needs to be reversed at the end: `collect_rev`)

**Comment by jflatow** [comment]
Closing WIP in favor of #287

**Pull request closed without merging**